sys/mount_linux.go

```diff
@@ -0,0 +1,139 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package sys
+
+import (
+	"runtime"
+	"syscall"
+	"unsafe"
+
+	"github.com/pkg/errors"
+)
+
+// FMountat performs mount from the provided directory.
+func FMountat(dirfd uintptr, source, target, fstype string, flags uintptr, data string) error {
+	var (
+		sourceP, targetP, fstypeP, dataP *byte
+		err                              error
+		n                                int
+		p                                [2]int
+		errno                            syscall.Errno
+	)
+
+	sourceP, err = syscall.BytePtrFromString(source)
+	if err != nil {
+		return err
+	}
+
+	targetP, err = syscall.BytePtrFromString(target)
+	if err != nil {
+		return err
+	}
+
+	fstypeP, err = syscall.BytePtrFromString(fstype)
+	if err != nil {
+		return err
+	}
+
+	if data != "" {
+		dataP, err = syscall.BytePtrFromString(data)
+		if err != nil {
+			return err
+		}
+	}
+
+	// open pipe to receive exit status from other thread
+	p[0], p[1] = -1, -1
+	if err = syscall.Pipe(p[:]); err != nil {
+		return err
+	}
+
+	defer func() {
+		syscall.Close(p[0])
+		syscall.Close(p[1])
+	}()
+
+	// NOTE: the runtime OS thread must be locked.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	if _, errno = forkAndMountat(dirfd,
+		uintptr(unsafe.Pointer(sourceP)),
+		uintptr(unsafe.Pointer(targetP)),
+		uintptr(unsafe.Pointer(fstypeP)),
+		flags,
+		uintptr(unsafe.Pointer(dataP)),
+		p[1]); errno != 0 {
+
+		return errors.Wrap(errno, "failed to fork thread")
+	}
+
+	// read exit status from child thread.
+	n, err = readlen(p[0], (*byte)(unsafe.Pointer(&errno)), int(unsafe.Sizeof(errno)))
```

Comment:
Is there no other way to do this without a raw syscall? That should only be necessary in the child process